### PR TITLE
Wrappers for lapack pstrf and pstf2, unit test included

### PR DESCRIPTION
**Pull request by tuxcell**

### CODE REVIEW

**Comment by tuxcell** [comment]
import numpy as np
from scipy.linalg import cholesky

''' R code
> ( m <- matrix(c(5,1,1,3),2,2) )
     [,1] [,2]
[1,]    5    1
[2,]    1    3
> ( cm <- chol(m) )
         [,1]      [,2]
[1,] 2.236068 0.4472136
[2,] 0.000000 1.6733201
> ( cm <- chol(m, pivot = TRUE) )
         [,1]      [,2]
[1,] 2.236068 0.4472136
[2,] 0.000000 1.6733201
attr(,"pivot")
[1] 1 2
attr(,"rank")
[1] 2'''

m = np.array([[5,1],[1,3]])
c=cholesky(m)
print(c)

[[2.23606798 0.4472136 ]
 [0.         1.67332005]]

cholesky(m, chol_pivot=True)

(array([[2.23606798, 0.4472136 ],
        [0.        , 1.67332005]]), 2, array([1, 2], dtype=int32))

''' R code
> ( m <- matrix(c(6,  3,  4,  8,3,  6,  5,  1, 4,  5, 10,  7,8,  1,  7, 25),4,4) )
     [,1] [,2] [,3] [,4]
[1,]    6    3    4    8
[2,]    3    6    5    1
[3,]    4    5   10    7
[4,]    8    1    7   25
> ( cm <- chol(m) )
        [,1]     [,2]     [,3]      [,4]
[1,] 2.44949 1.224745 1.632993  3.265986
[2,] 0.00000 2.121320 1.414214 -1.414214
[3,] 0.00000 0.000000 2.309401  1.587713
[4,] 0.00000 0.000000 0.000000  3.132491
> ( cm <- chol(m, pivot = TRUE) )
     [,1]     [,2]     [,3]      [,4]
[1,]    5 1.400000 0.200000 1.6000000
[2,]    0 2.835489 1.664616 0.6207041
[3,]    0 0.000000 1.785792 0.9221487
[4,]    0 0.000000 0.000000 1.4847115
attr(,"pivot")
[1] 4 3 2 1
attr(,"rank")
[1] 4'''

m = np.array([[6,  3,  4,  8],[3,  6,  5,  1],[4,  5, 10,  7],[8,  1,  7, 25]])
cholesky(m)

array([[ 2.44948974,  1.22474487,  1.63299316,  3.26598632],
       [ 0.        ,  2.12132034,  1.41421356, -1.41421356],
       [ 0.        ,  0.        ,  2.30940108,  1.58771324],
       [ 0.        ,  0.        ,  0.        ,  3.13249102]])

cholesky(m, chol_pivot=True)

(array([[5.        , 1.4       , 0.2       , 1.6       ],
        [0.        , 2.83548938, 1.66461565, 0.62070414],
        [0.        , 0.        , 1.78579246, 0.92214868],
        [0.        , 0.        , 0.        , 1.48471148]]),
 4,
 array([4, 3, 2, 1], dtype=int32))

File: scipy/linalg/decomp_cholesky.py
Comment (ilayn) [inline scipy/linalg/decomp_cholesky.py]: This switch must happen earlier so you don't call both potrf and pstrf depending on the flag. You need to solve only once. Either one or the other.

File: scipy/linalg/flapack_pos_semidef.pyf.src
Comment (ilayn) [inline scipy/linalg/flapack_pos_semidef.pyf.src]: This is much easier on the python side to cleanup and creates a maintenance burden here.

File: scipy/linalg/flapack_pos_semidef.pyf.src
Comment (ilayn) [inline scipy/linalg/flapack_pos_semidef.pyf.src]: `LDA >= max(1,shape(a, 0))`

File: scipy/linalg/flapack_pos_semidef.pyf.src
Comment (ilayn) [inline scipy/linalg/flapack_pos_semidef.pyf.src]: Why only prefix2? pstrf also has real and double flavors.

File: scipy/linalg/flapack_pos_semidef.pyf.src
Comment (ilayn) [inline scipy/linalg/flapack_pos_semidef.pyf.src]: This is much easier on the python side to cleanup and creates a maintenance burden here.

File: scipy/linalg/flapack_pos_semidef.pyf.src
Comment (ilayn) [inline scipy/linalg/flapack_pos_semidef.pyf.src]: `LDA >= max(1,shape(a, 0))`

**Review by ilayn — CHANGES_REQUESTED** [review]
Hi @tuxcell Thanks a lot for sparing the time for this. 

We have an important detail that, in my opinion, is a blocker (let's see what others will say); this routine doesn't check positive semidefiniteness and thus gives wrong results if the array is not PSD. And has a major call signature change that signals mixed messages to users.

I would prefer only wrapping  the LAPACK routines and not include the `cholesky` modifying parts. Then interested users can call themselves and others won't bother. As usual, we need to implement a number of tests for this in `test_lapack.py`.

**Comment by ilayn** [comment]
For the features you want to implement and to probe its applicability and scope, I would suggest opening an issue/asking it on the mailing list,  

Just for the future reference, please make separate branches to work on new features and to send PRs such that your standard working repo does not interfere with the PRs you have submitted. Once a PR is merged you can safely delete that branch and keep working on other branches.

File: scipy/linalg/flapack.pyf.src
Comment (tuxcell) [reply scipy/linalg/flapack.pyf.src]: Hi @ilayn,
Thank you so much for your feedback.
You improved the the code, made it more clear and provided some good insight about the request process.
Your suggestion of checking for PSD is very relevant. I will add that now.
About leaving the method "cholesky" alone, may I suggest adding a method "cholesky_full_pivot" with this feature?

Shall I remove 'flapack_pos_semidef.pyf.src' and add its contents to 'flapack_pos_def.pyf.src'?

File: scipy/linalg/decomp_cholesky.py
Comment (tuxcell) [reply scipy/linalg/decomp_cholesky.py]: I renamed full_pivot and pivot_tol but the default value of -1 is the same used by R's "chol", I couldn't find a reason for that choice yet, maybe it's a very commonly used value?

File: scipy/linalg/decomp_cholesky.py
Comment (tuxcell) [reply scipy/linalg/decomp_cholesky.py]: OK, fixed.

File: scipy/linalg/flapack_pos_semidef.pyf.src
Comment (tuxcell) [reply scipy/linalg/flapack_pos_semidef.pyf.src]: IMHO, cleaning code is best kept at the lowest level.

File: scipy/linalg/flapack_pos_semidef.pyf.src
Comment (tuxcell) [reply scipy/linalg/flapack_pos_semidef.pyf.src]: OK, done.

File: scipy/linalg/flapack_pos_semidef.pyf.src
Comment (tuxcell) [reply scipy/linalg/flapack_pos_semidef.pyf.src]: OK, done.

File: scipy/linalg/flapack_pos_semidef.pyf.src
Comment (tuxcell) [reply scipy/linalg/flapack_pos_semidef.pyf.src]: IMHO, cleaning code is best kept at the lowest level.

File: scipy/linalg/flapack_pos_semidef.pyf.src
Comment (tuxcell) [reply scipy/linalg/flapack_pos_semidef.pyf.src]: OK, done.

**Comment by tuxcell** [comment]
OK, thank you for the explanation.
I fixed the issues and added a few methods to test matrices for symmetry, PD, PSD, etc.
AFAIK, it looks fine now.

File: scipy/linalg/matfuncs.py
Comment (ilayn) [inline scipy/linalg/matfuncs.py]: These helper functions won't fly in their current state. Especially with hard coded tolerances. Also hermitianness is needed here. Real versions are not enough.

File: scipy/linalg/matfuncs.py
Comment (ilayn) [inline scipy/linalg/matfuncs.py]: every example code snippet needs to have `>>>` at the start of the line and then needs the results printed in the console. See other examples in the docs.

File: scipy/linalg/matfuncs.py
Comment (ilayn) [inline scipy/linalg/matfuncs.py]: empty line needed.

File: scipy/linalg/matfuncs.py
Comment (ilayn) [inline scipy/linalg/matfuncs.py]: I think symmetric matrices don't need any references.

File: scipy/linalg/matfuncs.py
Comment (ilayn) [inline scipy/linalg/matfuncs.py]: This would fail with big and small entries arrays. It needs to be tied to the 1- and inf-norm of the array entries.

File: scipy/linalg/matfuncs.py
Comment (ilayn) [inline scipy/linalg/matfuncs.py]: Same as above. If the array is too large it can always have some numerical noise around the small eigenvalues. Hence it is tricky for testing posdefness. That's why Cholesky decomposition is used for testing definiteness.

File: scipy/linalg/matfuncs.py
Comment (ilayn) [inline scipy/linalg/matfuncs.py]: Same as above.

File: scipy/linalg/matfuncs.py
Comment (ilayn) [inline scipy/linalg/matfuncs.py]: Empty lines.

File: scipy/linalg/matfuncs.py
Comment (ilayn) [inline scipy/linalg/matfuncs.py]: Same as above

File: scipy/linalg/matfuncs.py
Comment (ilayn) [inline scipy/linalg/matfuncs.py]: This will fail for negative definite matrices

File: scipy/linalg/flapack.pyf.src
Comment (ilayn) [reply scipy/linalg/flapack.pyf.src]: Yes that would be nice.

**Review by ilayn — CHANGES_REQUESTED** [review]
I would strongly suggest that we do only the LAPACK wrappers in this PR and leave the rest to mailing list discussion or an extra PR. Because it introduces significant changes to cholesky and also new functions that are not necessarily valid to put in.

File: scipy/linalg/matfuncs.py
Comment (tuxcell) [reply scipy/linalg/matfuncs.py]: Added hermitianness and symmetry tested with a new method.

File: scipy/linalg/matfuncs.py
Comment (tuxcell) [reply scipy/linalg/matfuncs.py]: OK, done

File: scipy/linalg/matfuncs.py
Comment (tuxcell) [reply scipy/linalg/matfuncs.py]: OK, done.

File: scipy/linalg/matfuncs.py
Comment (tuxcell) [reply scipy/linalg/matfuncs.py]: Removed references.

File: scipy/linalg/matfuncs.py
Comment (tuxcell) [reply scipy/linalg/matfuncs.py]: Replaced with new method.

File: scipy/linalg/matfuncs.py
Comment (tuxcell) [reply scipy/linalg/matfuncs.py]: OK, done.

File: scipy/linalg/matfuncs.py
Comment (tuxcell) [reply scipy/linalg/matfuncs.py]: Not anymore, now there is a tolerance argument.

**Comment by tuxcell** [comment]
Thank you for the help with improving the code, it is much better now.

**Comment by tuxcell** [comment]
import numpy as np

from scipy.linalg import is_matrix_hermitian,is_matrix_positive_definite, is_matrix_symmetric, is_matrix_positive_semidefinite,is_matrix_negative_definite,is_matrix_negative_semidefinite,is_matrix_indefinite

​

m = np.array([[5,1],[1,0]])

print(is_matrix_indefinite(m))

True

​

m = np.array([[-1,1],[1,-1]])

print(is_matrix_positive_definite(m))

print(is_matrix_positive_semidefinite(m))

print(is_matrix_negative_definite(m))

print(is_matrix_negative_semidefinite(m))

print(is_matrix_indefinite(m))

False
False
False
True
False

m = np.array([[5,1],[1,3]])

print(is_matrix_positive_definite(m))

print(is_matrix_positive_semidefinite(m))

print(is_matrix_negative_definite(m))

print(is_matrix_negative_semidefinite(m))

print(is_matrix_indefinite(m))

True
True
False
False
False

m = np.array([[1, -1, 0],[-1, 5, 0],[0, 0, 7]])

print(is_matrix_positive_definite(m))

print(is_matrix_positive_semidefinite(m))

print(is_matrix_negative_definite(m))

print(is_matrix_negative_semidefinite(m))

print(is_matrix_indefinite(m))

True
True
False
False
False

m = np.array([[1,-1],[-1,1]])

print(is_matrix_positive_definite(m))

print(is_matrix_positive_semidefinite(m))

print(is_matrix_negative_definite(m))

print(is_matrix_negative_semidefinite(m))

print(is_matrix_indefinite(m))

False
True
False
False
False

m = np.array([[-5,-1],[-1,-3]])

print(is_matrix_positive_definite(m))

print(is_matrix_positive_semidefinite(m))

print(is_matrix_negative_definite(m))

print(is_matrix_negative_semidefinite(m))

print(is_matrix_indefinite(m))

False
False
True
True
False

m = np.array([[-1,1],[1,-1]])

print(is_matrix_positive_definite(m))

print(is_matrix_positive_semidefinite(m))

print(is_matrix_negative_definite(m))

print(is_matrix_negative_semidefinite(m))

print(is_matrix_indefinite(m))

False
False
False
True
False

m = np.array([[5,1],[1,0]])

print(is_matrix_positive_definite(m))

print(is_matrix_positive_semidefinite(m))

print(is_matrix_negative_definite(m))

print(is_matrix_negative_semidefinite(m))

print(is_matrix_indefinite(m))

False
False
False
False
True

m = np.array([[1,2],[2,1]])

print(is_matrix_positive_definite(m))

print(is_matrix_positive_semidefinite(m))

print(is_matrix_negative_definite(m))

print(is_matrix_negative_semidefinite(m))

print(is_matrix_indefinite(m))

False
False
False
False
True

m = np.array([[-2,1,-2],[1,-2,1],[-2,1,-2]])

print(is_matrix_positive_definite(m))

print(is_matrix_positive_semidefinite(m))

print(is_matrix_negative_definite(m))

print(is_matrix_negative_semidefinite(m))

print(is_matrix_indefinite(m))

False
False
False
True
False

File: scipy/linalg/matfuncs.py
Comment (tuxcell) [reply scipy/linalg/matfuncs.py]: Cholesky decomposition for testing definiteness has issues with round off errors that eigenvalues can overcome. Both solutions are far from ideal, however they are simple to implement. 
"In order to perform Cholesky Decomposition of a matrix, the matrix has to be a positive definite matrix." And to test for definiteness, one uses Cholesky Decomposition. It sounds recursive and it does not test for semi positive or for negative definiteness, which can be done with eigenvalues.

Comment (ilayn) [reply scipy/linalg/matfuncs.py]: Semidefiniteness is much harder to test since it looks almost exact zeros.

File: scipy/linalg/matfuncs.py
Comment (ilayn) [inline scipy/linalg/matfuncs.py]: `-np.eye(3)` returns true but shouldn't

File: scipy/linalg/matfuncs.py
Comment (ilayn) [inline scipy/linalg/matfuncs.py]: `np.eye(3)` returns true but shouldn't

File: scipy/linalg/matfuncs.py
Comment (ilayn) [reply scipy/linalg/matfuncs.py]: Cholesky is cheaper to check but you are already eating away its advantage by computing eigenvalues which are more expensive.

Comment (tuxcell) [reply scipy/linalg/matfuncs.py]: You're right. So, I changed the code accordingly.

**Comment by ilayn** [comment]
Hi @tuxcell . Thanks for the updates but you are trying to include a numerically unstable algorithm which is the detection of semidefiniteness. It is not possible to detect via `?pstrf` routines as you are trying to include now and to be honest I don't think it is possible at all without a rank revealing QR or SVD (which both are quite expensive). 

This is the same message that the authors of the routine gave [in their report](http://www.netlib.org/lapack/lawnspdf/lawn161.pdf) with a nice example of an indefinite matrix marked as positive semidefinite. 

Just to let you make aware of these results and remind that it would be much better if we only include the wrappers and avoid the helper functions.

**Comment by tuxcell** [comment]
Hi @ilayn , thank you for the feedback. You are right, these are very expensive checks. I removed the helper functions, the changes in "cholesky" and left only the wrappers.
Thank you for  the report, it's really insightful.
Do I have to click on "close and comment" now?

File: scipy/linalg/matfuncs.py
Comment (tuxcell) [reply scipy/linalg/matfuncs.py]: It's been removed

**Comment by tuxcell** [comment]
> I would strongly suggest that we do only the LAPACK wrappers in this PR and leave the rest to mailing list discussion or an extra PR. Because it introduces significant changes to cholesky and also new functions that are not necessarily valid to put in.

OK, I removed the helper functions, the changes in "cholesky" and left only the wrappers.

**Comment by tuxcell** [comment]
Unit tests for both wrappers are included now.

**Comment by ilayn** [comment]
Some pep8 issues causing hiccups

**Comment by ilayn** [comment]
Hi @tuxcell I've tidied up a bit the wrappers and removed the cleaning parts since rank deficiency parts should also be zeroed out anyways and arguably better to do it by the user. Also that keeps the function signature compatible with native Fortran call. 

Just for the future reference, please make separate branches to work on new features and to send PRs such that your standard working repo does not interfere with the PRs you have submitted. Once a PR is merged you can safely delete that branch and keep working on other branches while keeping your master branch up to date with the official repo.

**Comment by tuxcell** [comment]
Hi @ilayn, this is great, now I can start a few branches for a few ideas to explore.
Thank you for the very positive communication, it's been a quite interesting learning process.

**Comment by ilayn** [comment]
My pleasure. Apparently the test I've included didn't go through. Let me quickly fix that and then we can hit the merge button.

**Comment by tuxcell** [comment]
Allright! Have a nice weekend! :)

**Comment by ilayn** [comment]
Ah this is why we need branches :) I'll revert the changes for a bit and then merge. Do you mind not adding changes to your master until we get it merged?

**Comment by tuxcell** [comment]
OK, I'll be reading and playing with branches in the mean time and leave master alone.

**Comment by ilayn** [comment]
OK all green and in it goes. Thanks @tuxcell !

**Pull request merged**
